### PR TITLE
8284813: x86 Code cleanup related to move instructions.

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Summary of changes:

- Correct feature checks in some assembler move instruction.
- Explicitly pass opmask register in routines accepting merge argument.
- Code re-organization related to move instruction, pull out the merge argument up to instruction pattern or top level caller.
- Add missing encoding based move elision checks in some macro assembly routines.

Kindly review and share your feedback.

Regards,
Jatin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8284813](https://bugs.openjdk.java.net/browse/JDK-8284813): x86 Code cleanup related to move instructions.

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [0792195e](https://git.openjdk.java.net/jdk/pull/8230/files/0792195e0770218d4e29fd1b81e9e736ae46c13f)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8230/head:pull/8230` \
`$ git checkout pull/8230`

Update a local copy of the PR: \
`$ git checkout pull/8230` \
`$ git pull https://git.openjdk.java.net/jdk pull/8230/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8230`

View PR using the GUI difftool: \
`$ git pr show -t 8230`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8230.diff">https://git.openjdk.java.net/jdk/pull/8230.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Good. Let me test it.

**Review by vnkozlov — COMMENTED** [review]
I found that Tobias already ran tier1-4 testing and it passed.
I am running tier1 with latest JDK.

You need second review. Could be from Intel.

**Review by vnkozlov — APPROVED** [review]
My tests passed.

File: src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp [hunk 1]
```diff
@@ -214,8 +214,8 @@ void MacroAssembler::copy32_masked_avx(Register dst, Register src, XMMRegister x
   mov64(temp, -1L);
   bzhiq(temp, temp, length);
   kmovql(mask, temp);
-  evmovdqu(type[shift], mask, xmm, Address(src, index, scale, offset), Assembler::AVX_256bit);
-  evmovdqu(type[shift], mask, Address(dst, index, scale, offset), xmm, Assembler::AVX_256bit);
+  evmovdqu(type[shift], mask, xmm, Address(src, index, scale, offset), true, Assembler::AVX_256bit);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp:217]: Should the merge parameter be set to false for load here?

File: src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp [hunk 2]
```diff
@@ -199,8 +199,8 @@ void MacroAssembler::copy64_masked_avx(Register dst, Register src, XMMRegister x
     mov64(temp, -1L);
     bzhiq(temp, temp, length);
     kmovql(mask, temp);
-    evmovdqu(type[shift], mask, xmm, Address(src, index, scale, offset), Assembler::AVX_512bit);
-    evmovdqu(type[shift], mask, Address(dst, index, scale, offset), xmm, Assembler::AVX_512bit);
+    evmovdqu(type[shift], mask, xmm, Address(src, index, scale, offset), true, Assembler::AVX_512bit);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp:202]: Should the merge parameter be set to false for load here?

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 3]
```diff
@@ -3021,53 +3021,27 @@ void Assembler::vmovdqu(Address dst, XMMRegister src) {
...
+  assert(VM_Version::supports_avx512vlbw(), "");
+  InstructionAttr attributes(vector_len, /* vex_w */ false, /* legacy_mode */ _legacy_mode_bw, /* no_mask_reg */ false, /* uses_vl */ true);
+  attributes.set_embedded_opmask_register_specifier(mask);
   attributes.set_is_evex_instruction();
   if (merge) {
     attributes.reset_is_clear_context();
   }
-  int prefix = (_legacy_mode_bw) ? VEX_SIMD_F2 : VEX_SIMD_F3;
-  int encode = vex_prefix_and_encode(dst->encoding(), 0, src->encoding(), (Assembler::VexSimdPrefix)prefix, VEX_OPCODE_0F, &attributes);
+  int encode = vex_prefix_and_encode(dst->encoding(), 0, src->encoding(), VEX_SIMD_F2, VEX_OPCODE_0F, &attributes);
```
Comment (sviswa7) [inline src/hotspot/cpu/x86/assembler_x86.cpp:3032]: The existing version (with no mask) was using VEX_SIMD_F2 or VEX_SIMD_F3 based on  avx512bw supported or not. With this change now the calling place need to be fixed. One place I see this being used is loadIotaIndices(). Please fix loadIotaIndices to use appropriate instruction for the platform. Is there any other place in array copy/clear?

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/assembler_x86.cpp:3032]: vector load operations in load_iota_indices are sensitive to vector length, a 64 byte iota values are loaded only for ByteVector.SPECIES_512 which necessitates existence of AVX512BW feature, I re-checked that copy/fill kernels  use appropriate instructions.

File: src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp (hunk 1 above)
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp:217]: DONE.

**Comment by jatin-bhateja** [comment]
Hi @sviswa7 , your comments have been addressed.

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 3 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/assembler_x86.cpp:3032]: Thanks for checking.

**Review by sviswa7 — APPROVED** [review]

**Pull request closed without merging**